File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/04-Contiguous-LinkedAllocation/LeetCode/LeetCode-0206-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    /// O(n) algorithm with O(1) extra space
    /// Avoids the if (r != null) check inside the while loop
    public ListNode reverseList(ListNode head) {
        ListNode p = null;
        ListNode q = head;
        ListNode r = head;

        while (q != null){
            r = r.next;
            q.next = p;
            p = q;
            q = r;
        } //end-while

        return p;              
    } // end-reverseList
}

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/04-Contiguous-LinkedAllocation/LeetCode/LeetCode-0206-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    # O(n) algorithm with O(1) extra space
    # Avoids the if (r != None) check inside the while loop
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        p = None
        q = head
        r = head

        while q != None:
            r = r.next
            q.next = p
            p = q
            q = r
        return p              
